internal/engine/upper.go

```diff
@@ -131,40 +136,57 @@ func (u Upper) CreateManifests(ctx context.Context, manifests []model.Manifest,
 			case <-ctx.Done():
 				return ctx.Err()
 			case event := <-sw.events:
-				oldState := buildStates[event.manifest.Name]
-				buildState := oldState.NewStateWithFilesChanged(event.files)
-				buildStates[event.manifest.Name] = buildState
-
-				spurious, err := buildState.OnlySpuriousChanges()
-				if err != nil {
-					logger.Get(ctx).Infof("build watch error: %v", err)
-				}
+				if eventContainsConfigFiles(event) {
+					newManifest, err := getNewManifestFromTiltfile(ctx, event.manifest.Name)
+					if err != nil {
+						logger.Get(ctx).Infof("getting new manifest error: %v", err)
+						haveValidTiltfile = false
+						continue
+					}
+					buildState := BuildStateClean
+					err = u.buildManifestFromBuildState(ctx, newManifest, buildState)
+					if err != nil {
+						logger.Get(ctx).Infof("config watch error: %v", err)
+						haveValidTiltfile = false
+						continue
+					}
+					haveValidTiltfile = true
+					u.manifestOverlay[newManifest.Name] = newManifest
+				} else {
+					oldState := u.buildStates[event.manifest.Name]
+					buildState := oldState.NewStateWithFilesChanged(event.files)
+					u.buildStates[event.manifest.Name] = buildState
 
-				if spurious {
-					// TODO(nick): I think we probably want to log when this happens?
-					continue
-				}
+					spurious, err := buildState.OnlySpuriousChanges()
+					if err != nil {
+						logger.Get(ctx).Infof("build watch error: %v", err)
+					}
 
-				u.logBuildEvent(ctx, event.manifest, buildState)
+					if spurious {
+						// TODO(nick): I think we probably want to log when this happens?
+						continue
+					}
 
-				result, err := u.b.BuildAndDeploy(
-					ctx,
-					event.manifest,
-					buildState)
-				if err != nil {
-					if isPermanentError(err) {
+					manifest := u.getMostRecentManifest(event)
+					if !haveValidTiltfile {
```

Comment:
I'm pretty confident that this `if` statement is unnecessary? if the Tiltfile is broken, and no config files have changed, then it will still be broken. if you remove this block, do any tests break?

Reply:
Yep, good call.